[quant][graphmode] Add fusion for quantized linear

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25974 [quant][graphmode] use whitelist for selecting observed values
* #25625 [quant][graphmode] Fold quantize op into module
* **#25624 [quant][graphmode] Add fusion for quantized linear**
* #25623 Port fuse_linear from pytorch/tvm
* #25505 [quant][graphmode] Support quantizing any methods called

Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208891](https://our.internmc.facebook.com/intern/diff/D17208891)

jerryzh168 committed:
[quant][graphmode] Add fusion for quantized linear

Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208891](https://our.internmc.facebook.com/intern/diff/D17208891)

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208891](https://our.internmc.facebook.com/intern/diff/D17208891)

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208891](https://our.internmc.facebook.com/intern/diff/D17208891)

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208891](https://our.internmc.facebook.com/intern/diff/D17208891)

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208891](https://our.internmc.facebook.com/intern/diff/D17208891)

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208891](https://our.internmc.facebook.com/intern/diff/D17208891)

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -629,10 +633,39 @@ graph(%a_quant, %w_quant, %b_quant, %a_scale, %a_zero_point, %a_dtype, %w_scale,
+  };
+  for (const auto& item: pattern_and_replacements) {
+    SubgraphRewriter rewriter;
+    rewriter.RegisterRewritePattern(item.first, item.second);
```
As in the previous PR, you could register all the patterns together and run the rewriter once.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,6 +1,7 @@
 #include <torch/csrc/jit/passes/quantization.h>
 #include <torch/csrc/jit/passes/subgraph_rewrite.h>
 #include <torch/csrc/jit/passes/constant_propagation.h>
+#include <torch/csrc/jit/passes/fuse_linear.h>
```
I suspect clang-format would like to rearrange the headers in the alphabetical order - please run clang-format on all changes (the constant_propagation header added in the previous PR should be reformatted in the corresponding PR too).

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -629,10 +633,39 @@ graph(%a_quant, %w_quant, %b_quant, %a_scale, %a_zero_point, %a_dtype, %w_scale,
+        %2 : int = prim::Constant[value=2]()
+        %3 : int = prim::Constant[value=3]()
+        %in_param : int[] = prim::ListConstruct(%0, %2, %3, %1)
+        %a_perm : Tensor = aten::permute(%a_quant, %in_param)
```
If we have more patterns in future, it might make sense to hoist them to a separate file and just include it here. `quantization.cpp` becomes quite large.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -604,7 +605,11 @@ void FoldQuantNodesIntoInputsOutputs(std::shared_ptr<Graph>& graph) {
 
 void QuantFusion(std::shared_ptr<Graph>& graph) {
-  std::string pattern = R"(
+  // First fuse aten::linear op
```
Could we list all the patterns we're rewriting in a comment (in a more compact way then in the actual pattern)? Something like
```
q(linear(dq(a), dq(w), dq(b))) --> to_nchw(fbgemm_linear(prepack(to_nhwc(a)),
                                                         prepack(to_nhwc(w)),
                                                         prepack(to_nhwc(b))))
```

ZolotukhinM reviewed (approved):
👍

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208891](https://our.internmc.facebook.com/intern/diff/D17208891)

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208891](https://our.internmc.facebook.com/intern/diff/D17208891)

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208891](https://our.internmc.facebook.com/intern/diff/D17208891)

jerryzh168 committed:
Update on "[quant][graphmode] Add fusion for quantized linear"


Summary:
First fuse the splitted op into aten::linear and then fuse
`dequant - aten::linear - quant` into quantized linear op

Test Plan:
python test/test_jit.py 'TestJit.quant_fusion'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208891](https://our.internmc.facebook.com/intern/diff/D17208891)
